mlir/include/mlir/Conversion/LLVMCommon/TypeConverter.h

```diff
@@ -258,6 +259,9 @@ class LLVMTypeConverter : public TypeConverter {
   /// Convert a 1D vector type into an LLVM vector type.
   FailureOr<Type> convertVectorType(VectorType type) const;
 
+  /// Convert AMX tile type x86_amx type.
+  Type convertAMXTileType(amx::TileType type) const;
```

Comment:
Hm, I see the chicken and egg problem here. Technically, the LLVM header should not include an AMX header, but the conversion is on an AMX type.

Perhaps the AMX tile type needs to be "lowered" to a vector type before it gets converted to LLVM's `X86_amx` type, but that's just odd.

Alternatively, since the `X86_amx` type is actually in LLVM itself, the type could be declared there, then you don't need to cross-include a higher level file.

Reply:
To be clear, you may need to declare the type twice: `amx::TileType` that lowers directly to `llvm::x86_amx` without need for conversion.

Reply:
I don't see how I can go directly from `amx::TileType` to `llvm::x86_type`. For proper conversion of MLIR module to LLVM module, all types should be LLVM compatible and this is defined by LLVMIR dialect. I cannot declare an external LLVM compatible type.

The only simple way to get rid of the new dependency here is to move `amx::TileType` to bulitin types. It would better match what we have in LLVM, but doesn't feel good.

Probably, the best solution would be to introduce a type interface for converting custom types to LLVMIR dialect types. Then we can utilize this interface in `LLVMTypeConverter` and keep `amx::TileType` to `LLVMX86AMXType` conversion code in AMX dialect removing dependency of LLVM converter on AMX dialect.

Reply:
> I cannot declare an external LLVM compatible type.

I'm not sure what you mean by this. Should be possible to create a new type in the LLVM dialect that lowers directly to the LLVM IR type.

> The only simple way to get rid of the new dependency here is to move `amx::TileType` to bulitin types.

That's not the right way around, indeed.

> Probably, the best solution would be to introduce a type interface for converting custom types to LLVMIR dialect types.

Is this the only case of custom type conversion into LLVM? Are there no other examples in tree?

Reply:
> I'm not sure what you mean by this. Should be possible to create a new type in the LLVM dialect that lowers directly to the LLVM IR type.

This patch already has such a type in the LLVM dialect. It's `LLVM::LLVMX86AMXType` added to `LLVMTypes.td`. Changes added to the LLVM type converter are required to convert `amx::TileType` to this `LLVMX86AMXType`. These changes allow existing passes (such as CF to LLVM lowering) to properly convert `amx::TileType` values to `LLVM::LLVMX86AMXType` values.

> Is this the only case of custom type conversion into LLVM? Are there no other examples in tree?

I didn't see any similar cases. Users have to use their type converters to handle non-builtin types.